pkg/util/strategicpatch/patch.go

```diff
@@ -1447,3 +1453,34 @@ func GetServerSupportedSMPatchVersion(discoveryClient discovery.DiscoveryInterfa
 	}
 	return Unknown, fmt.Errorf("The version is too old: %v\n", serverVersion)
 }
+
+// GuidedJsonMerge merges two JSON formatted items by following the structure of dataStruct's type. We use
+// dataStruct's type to respect strategicMerge guidelines and ensure the structure of the JSON items
+// (e.g. no fields are missing, no additional fields exist, etc)
+func GuidedJsonMerge(j1, j2 []byte, dataStruct interface{}) ([]byte, error) {
+	j1Map := map[string]interface{}{}
+	if len(j1) > 0 {
+		if err := json.Unmarshal(j1, &j1Map); err != nil {
+			return nil, fmt.Errorf("Invalid JSON document")
+		}
+	}
+
+	j2Map := map[string]interface{}{}
+	if len(j2) > 0 {
+		if err := json.Unmarshal(j2, &j2Map); err != nil {
+			return nil, fmt.Errorf("Invalid JSON document")
```

Comment:
Including the err in the message would be useful.  e.g. fmt.Errorf("... %v", err)